File: r/src/test/scala/org/apache/zeppelin/rinterpreter/RInterpreterTest.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.zeppelin.rinterpreter

import java.util.Properties

import org.apache.zeppelin.RTest
import org.apache.zeppelin.interpreter.{Interpreter, InterpreterContext, InterpreterResult, InterpreterGroup}
import org.scalatest.Matchers._
import org.scalatest._
import java.util.ArrayList

class RInterpreterTest extends FlatSpec {

  RContext.resetRcon()

  class RIntTester extends RInterpreter(new Properties(), startSpark = false) {

    def interpret(s: String, interpreterContext: InterpreterContext): InterpreterResult = {
      val result : Array[String] = rContext.evalS1(s)
      new InterpreterResult(InterpreterResult.Code.SUCCESS, result.mkString("\n"))
    }
  }
  val rint = new RIntTester()

  "An RInterpreter" should "exist" in {
    assert(rint != null)
  }

  it should "not complain when we assign it a group" in {
    val grp : InterpreterGroup = new InterpreterGroup("test")
    val lst : ArrayList[Interpreter] = new ArrayList[Interpreter]()
    lst.add(rint)
    grp.put(rint.getClassName(), lst)
    rint.setInterpreterGroup(grp)
  }

  it should "create a fresh rContext when we ask for one" in {
    assert(! rint.getrContext.isOpen)
  }

  it should "open"  taggedAs(RTest) in {
    rint.open()
    assert(rint.getrContext.isOpen)
  }

  it should "have rzeppelin available"  taggedAs(RTest) in {
    assume(rint.getrContext.isOpen)
    assert(rint.getrContext.testRPackage("rzeppelin"))
  }
  it should "have an rContext able to do simple addition" taggedAs(RTest)  in {
    assume(rint.getrContext.isOpen)
    assert(rint.getrContext.evalI0("2 + 2") == 4)
  }



/*  it should "have a functional completion function" taggedAs(RTest) in {
    val result = rint.hiddenCompletion("hi", 3)
    result should (contain ("hist"))
  }*/

  it should "have a working progress meter" in {
    rint.getrContext.setProgress(50)
    assertResult(50) {
      rint.getrContext.getProgress
    }
  }

  it should "have persistent properties" in {
    val props = new Properties()
    props.setProperty("hello", "world")
    rint.setProperty(props)
    assertResult("world") {
      rint.getProperty("hello")
    }
  }

  var rint2 : RIntTester = null

  it should "Share RContexts if they share the same InterpreterGroup" in {
    rint2 = new RIntTester()
    val lst : ArrayList[Interpreter] = new ArrayList[Interpreter]()
    lst.add(rint2)
    val grp = rint.getInterpreterGroup()
    grp.put(rint2.getClassName(), lst)
    rint2.setInterpreterGroup(grp)
    rint2.open()
    rint.getrContext should be theSameInstanceAs rint2.getrContext
  }

  "Opening the second RInterpreter" should "not have closed the first RContext" in {
    assert(rint.getrContext.isOpen)
  }

  var rint3 : RIntTester = null

  "An RInterpreter in a different InterpreterGroup" should "have a different R Context" in {
    rint3 = new RIntTester()
    val grp : InterpreterGroup = new InterpreterGroup("othertest")
    val lst : ArrayList[Interpreter] = new ArrayList[Interpreter]()
    lst.add(rint3)
    grp.put(rint3.getClassName(), lst)
    rint3.setInterpreterGroup(grp)
    rint3.open()
    rint3.getrContext shouldNot be theSameInstanceAs rint2.getrContext
  }

  "The first RInterpreter" should "close politely" in {
    rint.close()
    assert(!rint.getrContext.isOpen)
  }

  "and so" should "the other one" in {
    rint2.close()
    assert(!rint2.getrContext.isOpen)
  }

  "and " should "the third one" in {
    rint3.close()
    assert(!rint2.getrContext.isOpen)
  }

//  fixture.sparky.close()

}

File: rlang/src/test/java/org/apache/zeppelin/r/RInterpreterTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


package org.apache.zeppelin.r;

import org.apache.commons.lang3.exception.ExceptionUtils;
import org.apache.zeppelin.interpreter.Interpreter;
import org.apache.zeppelin.interpreter.InterpreterContext;
import org.apache.zeppelin.interpreter.InterpreterException;
import org.apache.zeppelin.interpreter.InterpreterGroup;
import org.apache.zeppelin.interpreter.InterpreterOutput;
import org.apache.zeppelin.interpreter.InterpreterResult;
import org.apache.zeppelin.interpreter.LazyOpenInterpreter;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.io.IOException;
import java.util.HashMap;
import java.util.Properties;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

public class RInterpreterTest {

  private RInterpreter rInterpreter;

  @Before
  public void setUp() throws InterpreterException {
    Properties properties = new Properties();
    properties.setProperty("zeppelin.R.knitr", "true");
    properties.setProperty("spark.r.backendConnectionTimeout", "10");

    InterpreterContext context = getInterpreterContext();
    InterpreterContext.set(context);
    rInterpreter = new RInterpreter(properties);

    InterpreterGroup interpreterGroup = new InterpreterGroup();
    interpreterGroup.addInterpreterToSession(new LazyOpenInterpreter(rInterpreter), "session_1");
    rInterpreter.setInterpreterGroup(interpreterGroup);

    rInterpreter.open();
  }

  @After
  public void tearDown() throws InterpreterException {
    rInterpreter.close();
  }

  @Test
  public void testSparkRInterpreter() throws InterpreterException, InterruptedException, IOException {
    InterpreterResult result = rInterpreter.interpret("1+1", getInterpreterContext());
    assertEquals(InterpreterResult.Code.SUCCESS, result.code());
    assertTrue(result.message().get(0).getData().contains("2"));

    InterpreterContext context = getInterpreterContext();
    result = rInterpreter.interpret("foo <- TRUE\n" +
            "print(foo)\n" +
            "bare <- c(1, 2.5, 4)\n" +
            "print(bare)\n" +
            "double <- 15.0\n" +
            "print(double)", context);
    assertEquals(InterpreterResult.Code.SUCCESS, result.code());
    assertTrue(result.toString(),
                       result.message().get(0).getData().contains("[1] TRUE\n" +
                               "[1] 1.0 2.5 4.0\n" +
                               "[1] 15\n"));

    // plotting
    context = getInterpreterContext();
    context.getLocalProperties().put("imageWidth", "100");
    result = rInterpreter.interpret("hist(mtcars$mpg)", context);
    assertEquals(InterpreterResult.Code.SUCCESS, result.code());
    assertEquals(1, result.message().size());
    assertEquals(InterpreterResult.Type.HTML, result.message().get(0).getType());
    assertTrue(result.message().get(0).getData().contains("<img src="));
    assertTrue(result.message().get(0).getData().contains("width=\"100\""));

    result = rInterpreter.interpret("library(ggplot2)\n" +
            "ggplot(diamonds, aes(x=carat, y=price, color=cut)) + geom_point()", getInterpreterContext());
    assertEquals(InterpreterResult.Code.SUCCESS, result.code());
    assertEquals(1, result.message().size());
    assertEquals(InterpreterResult.Type.HTML, result.message().get(0).getType());
    assertTrue(result.message().get(0).getData().contains("<img src="));

    // sparkr backend would be timeout after 10 seconds
    Thread.sleep(15 * 1000);
    result = rInterpreter.interpret("1+1", getInterpreterContext());
    assertEquals(InterpreterResult.Code.ERROR, result.code());
    assertTrue(result.message().get(0).getData().contains("sparkR backend is dead"));
  }

  @Test
  public void testInvalidR() throws InterpreterException {
    tearDown();

    Properties properties = new Properties();
    properties.setProperty("zeppelin.R.cmd", "invalid_r");
    properties.setProperty("spark.master", "local");
    properties.setProperty("spark.app.name", "test");

    InterpreterGroup interpreterGroup = new InterpreterGroup();
    Interpreter rInterpreter = new LazyOpenInterpreter(new RInterpreter(properties));
    interpreterGroup.addInterpreterToSession(rInterpreter, "session_1");
    rInterpreter.setInterpreterGroup(interpreterGroup);

    InterpreterContext context = getInterpreterContext();
    InterpreterContext.set(context);

    try {
      rInterpreter.interpret("1+1", getInterpreterContext());
      fail("Should fail to open SparkRInterpreter");
    } catch (InterpreterException e) {
      String stacktrace = ExceptionUtils.getStackTrace(e);
      assertTrue(stacktrace, stacktrace.contains("No such file or directory"));
    }
  }

  private InterpreterContext getInterpreterContext() {
    InterpreterContext context = InterpreterContext.builder()
            .setNoteId("note_1")
            .setParagraphId("paragraph_1")
            .setInterpreterOut(new InterpreterOutput(null))
            .setLocalProperties(new HashMap<>())
            .build();
    return context;
  }
}
